daemon/container_operations.go

```diff
@@ -1069,13 +1069,27 @@ func (daemon *Daemon) ConnectToNetwork(ctx context.Context, ctr *container.Conta
 		}
 
 		n, err := daemon.FindNetwork(idOrName)
-		if err == nil && n != nil {
+		if err == nil {
+			if _, ok := ctr.NetworkSettings.Networks[idOrName]; ok {
+				return fmt.Errorf("endpoint with name %s already exists in network %s", strings.TrimPrefix(ctr.Name, "/"), idOrName)
+			}
+
 			if err := daemon.updateNetworkConfig(ctr, n, endpointConfig, true); err != nil {
```

Comment:
Looking at this code, I see it's calling `Daemon.updateNetworkConfig()` https://github.com/moby/moby/blob/6ac445c42badc98b416baa32cca5d6f8b712b359/daemon/container_operations.go#L658

Which performs some validation, then (if `updateSettings=true`) calls `Daemon.updateNetworkSettings()` https://github.com/moby/moby/blob/6ac445c42badc98b416baa32cca5d6f8b712b359/daemon/container_operations.go#L220

Which ... looks to ignore duplicates; https://github.com/moby/moby/blob/6ac445c42badc98b416baa32cca5d6f8b712b359/daemon/container_operations.go#L246-L247


My train of thought here is;

- Are we doing too much here, because `Daemon.updateNetworkConfig()` / `Daemon.updateNetworkSettings()` also lookup the network, so even curious if the second branch below should be handed there.
- Should `Daemon.updateNetworkSettings()` itself bail out early if the container is already attached?
- :point_up: are there situations where it's allowed to attach to a network, but with updated configuration?


@akerouanton @robmry any thoughts?

Reply:
@thaJeztah 

Yes, you’re right. Actually, my opinion is that when a container is not started or is stopped, if we try to connect a network that is already connected, it would be better to return an error instead of updating, just like when the container is already started.

It would be better if the update could happen in another route (if possible). This might be a bit confusing. Also, we can connect a network to a container even if that network does not exist at all.

Reply:
> if we try to connect a network that is already connected, it would be better to return an error instead of updating, just like when the container is already started.

Agreed. I guess I was mostly wondering if there'd he situations where it's possible to update network-connection settings, but don't think we have this, and for swarm containers, those would be re-created.

> It would be better if the update could happen in another route (if possible). This might be a bit confusing. Also, we can connect a network to a container even if that network does not exist at all.

Yeah; for full transparency, (lib)network code has made some odd design choices in the past where errors were logged, but otherwise ignored. Some of those were for backward-compatibility, others were related to (e.g.) swarm networks, where networks are dynamically rolled out to nodes, and may not be available _yet_ in certain stages. We're trying to get rid of some of those code-paths (fail loud instead of ignore silently), but many of those design choices have not been documented, so it sometimes "two steps forward, one step back", and with a large and diverse user-base, we always discover "someone, somewhere" depending on some of these :joy:

Reply:
@thaJeztah 

Do you think applying these changes will cause any issues? Are we going to lose our backward compatibility?
From a quick look at Docker APIs, I didn’t see anything particularly concerning.

Reply:
It's bound to break a workflow somewhere! The question is whether it's worth that hit to get the errors out earlier. I'm not sure, perhaps it is.

Trying to dig into why the code is the way it is, not much luck ...

- https://github.com/moby/moby/pull/26287
  - made it possible to add networks to a stopped container on Windows, but it only moved the code. I don't think there's any reason this change would introduce a problem there.
- https://github.com/moby/moby/pull/25962
  - is where the code came from, so it was related to Swarm. But, it's not clear from the PR/commit whether there's a need for a container to accept an attachment for a network it doesn't know about yet (maybe on a specific worker node, I'm not sure - it needs further investigation/testing, or input from someone who knows Swarm better than I do).

Even with this change, a container can be created with a connection to a nonexistent network (`docker create --network nosuchnet ...`). So, creating a new container would become different to connecting a network to one that's not started yet, or that's been stopped... I think those should remain consistent, one way or the other.

Reply:
> * it needs further investigation/testing

what kind of test do we need to run?

> So, creating a new container would become different to connecting a network to one that's not started yet, or that's been stopped... I think those should remain consistent, one way or the other.

Yes, that's correct. I had considered this in relation to `docker network connect` and hadn’t paid attention to the point you mentioned.